### PR TITLE
front: fix slopes editor form validation

### LINKED ISSUE #3479: Editor : slopes with no gradient values can be save
### What happened?

When editing a slope, a message says that gradient is a required property. Despite this, it is possible to save modifications without defing the gradient value.

Image

### What did you expect to happen?

We should block the save untill the user defines gradient value

### How can we reproduce it (as minimally and precisely as possible)?

1. Go to the editor module
2. Select a track and click on the edition button
3. Click to edit the slopes without defining the gradient
4. Click on the save button. A message saying "the modifications are saved" appears. 

### What operating system, browser and environment are you using?
- **Browser**: Chrome v110
- **OS**: Windows 10
- **Env**: Dgex dev

### OSRD version (top right corner `Account` button > `Informations`)

5d292539

### PR DESCRIPTION
**Pull request by Wadjetz**
close #3479 
- [x] Make the field optional
- [x] Graph modifications
Capture d’écran 2023-11-15 à 17 47 20

No value
Capture d’écran 2023-11-15 à 17 47 33

Zero 
Capture d’écran 2023-11-15 à 17 52 49
Capture d’écran 2023-11-17 à 09 27 22

### CODE REVIEW

**Review by SharglutDev — APPROVED** [review]
Good job for the catch of this customOnChange ! 

Tested in local and everything seems to be working as expected. Tried also the forms for signals, detectors, switches, buffer stops and seen nothing strange.

**Review by clarani — APPROVED** [review]
LGTM :white_check_mark:

**Review by clarani — CHANGES_REQUESTED** [review]
We need to wait for Giuliana's point of view

**Review by SharglutDev — CHANGES_REQUESTED** [review]
Take you for the changes. As checked with you, I think the curves and slopes values should be set to 0 as soon as we enter in the form.

Otherwise, I think it's strange that if the value is at 0, we see a `-` when hovering the interval editor. It might mislead the user as it's the same behavior than the `undefined` one.

image

Also you have some warnings in the file

**Review by SharglutDev — CHANGES_REQUESTED** [review]
Thank you for the changes.
When I try to create a new tracksection, after drawing it and fill the required inputs, doesn't work. Its saying that curves and slopes are required properties even though gradiant and radius are set to 0 when entering the form.

image

On the other hand, the behavior when editing a track section with unspecified gradiants or radius is great, they are set automatically to 0, good job !

File: front/src/common/IntervalsDataViz/IntervalItem.tsx [hunk 1]
```diff
@@ -56,21 +56,25 @@ const IntervalItem = <T extends { [key: string]: string | number }>({
       valueText = `${interval[field]}`;
     }
   }
+
+  const isNoData =
```
Comment (SharglutDev) [inline front/src/common/IntervalsDataViz/IntervalItem.tsx]:
```suggestion
  const hasNoData =
```

File: front/src/common/IntervalsDataViz/IntervalItem.tsx [hunk 2]
```diff
@@ -56,21 +56,25 @@ const IntervalItem = <T extends { [key: string]: string | number }>({
...
+    !!field &&
+    (segment === undefined || segment[field] === undefined || segment[field] === emptyValue);
+  const isHasData =
```
Comment (SharglutDev) [inline front/src/common/IntervalsDataViz/IntervalItem.tsx]:
```suggestion
  const hasData =
```

File: front/src/common/IntervalsDataViz/IntervalItem.tsx [hunk 3]
```diff
@@ -56,21 +56,25 @@ const IntervalItem = <T extends { [key: string]: string | number }>({
...
-          segment[field] !== undefined &&
-          segment[field] !== emptyValue &&
-          'with-data',
-        field &&
-          (segment === undefined ||
-            segment[field] === undefined ||
-            segment[field] === emptyValue) &&
-          'no-data',
+        isHasData && 'with-data',
+        isNoData && 'no-data',
```
Comment (SharglutDev) [inline front/src/common/IntervalsDataViz/IntervalItem.tsx]: We have a convention when using condition in cx
```suggestion
        { 
          'with-data': isHasData,
          'no-data': isNoData,
        },
```

**Review by SharglutDev — APPROVED** [review]
Great job for this PR, thank you ! Tested and everything works as expected.

Left some naming comments. As I will be away the next 2 business days, I'll let @clarani resolve those so you are not blocked to merge the PR.

File: front/src/common/IntervalsDataViz/data.ts [hunk 4]
```diff
@@ -688,8 +688,10 @@ export function entityDoUpdate<T extends EditorEntity>(entity: T, sourceLine: Li
 export function getFieldJsonSchema(
   fieldSchema: JSONSchema7,
   rootSchema: JSONSchema7,
-  enhancement: { [key: string]: JSONSchema7Definition } = {}
+  enhancement: { [key: string]: JSONSchema7Definition },
+  requiredFilter?: (required: string[]) => string[]
 ): JSONSchema7 {
+  const enhancementValues = enhancement ?? {};
```
Comment (clarani) [inline front/src/common/IntervalsDataViz/data.ts]: Why did you do that ? I find it clearer to have the default value in the definition of the function (see before changes)
```suggestion
```

File: front/src/applications/editor/components/EditorForm.tsx [hunk 5]
```diff
@@ -84,6 +84,16 @@ function EditorForm<T extends Omit<EditorEntity, 'objType'> & { objType: string
...
 
+  useEffect(() => {
+    if (!data.properties.curves || !data.properties.slopes) {
+      setFormData({
+        ...data.properties,
+        slopes: data.properties.slopes ?? [],
+        curves: data.properties.curves ?? [],
+      });
+    }
+  }, [data.properties.length]);
```
Comment (clarani) [inline front/src/applications/editor/components/EditorForm.tsx]: You don't need this, the best thing to do I think is to initialize curves and slopes to [] in getInitialState (see getNewLine, add curves: [] and slopes: []).
You will need to add curves and slopes to the TrackSectionEntity
```
      slopes: LinearMetadataItem<{ gradient: number }>[];
      curves: LinearMetadataItem<{ radius: number }>[];
```
So you can remove the useEffect
```suggestion
```

**Review by clarani — CHANGES_REQUESTED** [review]
Thanks for this PR, almost there :)

File: front/src/common/IntervalsDataViz/data.ts (hunk 4 above)
Comment (Wadjetz) [reply front/src/common/IntervalsDataViz/data.ts]: The linter require params with default values to be the last

**Review by clarani — APPROVED** [review]
LGTM and tested ! Nice job, thank you very much :pray: :tada:

**Pull request merged**